### ISSUE TITLE
osu!lazer won't display CJK characters in China style.

### ISSUE THREAD

**Issue by SirnoChan** [ppy/osu#13843]
**Describe the bug:** 
I tried apply Chinese l10n to new version osu!lazer, and my Windows 10 system is set up with China locale/region. After recent updates, my osu!lazer instance starts supporting i18n translations as well. But this brings up a problem which may be not a rare occasion in similar circumstances. osu!lazer doesn't display Chinese characters in Chinese style but in Japanese style. CJK Unified Ideographs have some characters of common unicode codes, but the writtings are different, and if software doesn't recognise users' locale and doesn't deal with them properly, users will see a writting style they are not fimilar to.

Take the below screenshot as an example. In the screenshot, the "置" in "设置" (settings) should be displayed as !["置" in China locale style](http://zht.glyphwiki.org/glyph/u7f6e-g.50px.png) but it shows !["置" is Japan locale style](http://zht.glyphwiki.org/glyph/u7f6e-j.50px.png), and the "改" in "改变" (to change) should be displayed as !["改" in China locale style](http://zht.glyphwiki.org/glyph/u6539-g.50px.png) but it shows !["改" in Japan locale style](http://zht.glyphwiki.org/glyph/u6539-j.50px.png).
**Screenshots or videos showing encountered issue:** 
![screenshot](https://user-images.githubusercontent.com/49866669/125060625-23547b00-e0df-11eb-8c10-01b82302d9ac.png)

**osu!lazer version:** 
2021.707.0-lazer
**Logs:**
Not applicable.

**Comment by gagahpangeran** [comment]
The string is from this localisation.

https://github.com/ppy/osu-resources/blob/2ba3c03018270c7f99fdbfa9d20775e326b3315d/osu.Game.Resources/Localisation/Settings.zh.resx#L61-L66

If you want to correct it, please contribute and change it in [crowdin](https://crowdin.com/project/osu-web) and the localisation will be updated frequenly from that.

**Comment by bdach** [comment]
This is not a problem with the localisation. The unicode standard (namely the unified ideographs section, as stated in the OP), allows a single Unicode codepoint to be displayed in a different manner depending on the language in which the codepoint is being rendered.

I'd recommend checking out the relevant section of the unicode spec and looking at the `0x7F6E` codepoint, which is one of the ones mentioned in the OP.

**Comment by gagahpangeran** [comment]
Ah, I totally missed about that.

Thanks for the correction.

**Comment by kj415j45** [comment]
Ref: https://github.com/ppy/osu/discussions/18522

Doing some search in repos and I'm sure this issue can be fixed at lazer side (osu and/or osu-resource), rather than osu-framework:
- Allow user to select font that used in lazer from system.
- Load and use Noto Sans SC/TC (only) when ui language is Chinese.

**Comment by smoogipoo** [comment]
It will need support from osu!framework to clear the glyph/texture atlases.

**Comment by peppy** [comment]
I'm 99% sure that isn't the solution we want, as it would display incorrect characters in metadata when the language of the metadata doesn't match the user's language. Maybe broken in a better way than it is now, but still an issue to consider.

As mentioned by @bdach this should be localised to usages rather than applied on a game-wide level. Maybe something which is added to `LocalisedString`.

**Comment by kj415j45** [comment]
> I'm 99% sure that isn't the solution we want, as it would display incorrect characters in metadata when the language of the metadata doesn't match the user's language. Maybe broken in a better way than it is now, but still an issue to consider.

Yes, but characters in metadata is quite a different problem. It should match the music's language. We may need to add a property so that these metadata can be rendered with right font. Or directly use `language` defined in mp3 or beatmapset.
However, those strings used by UI should strictly render with user's locale/language settings. That's the main problem in this issue.

---

> It will need support from osu!framework to clear the glyph/texture atlases.

Yes, load resources without any usage is a bad smell, but how much memory will these unused glyph/texture waste? Is it worth to write code to unload them?

**Comment by Wanye-7300** [comment]
Excuse me, is there any progress on this issue? Thanks a lot.

**Comment by peppy** [comment]
There's no progress. Considering stable never got this right, I wouldn't have high hopes on this being fixed anytime soon. It's not a simple one.

**Comment by cihe13375** [comment]
> Considering stable never got this right

Even though this is not 100% correct on stable either, I want to point out that stable deals with it way better:
1) the UI display is correct. The 1st image below is in simplified Chinese, and the 2nd image is in Japanese. Note the character 置.

![Image](https://github.com/user-attachments/assets/4020d3c6-2243-4037-a3c6-2039b14cfe23)
![Image](https://github.com/user-attachments/assets/4d3b5f36-3cdb-4bd9-81a6-cdbcc368f81b)

2) regarding the metadata, I think the stable logic is the following:
a) if the field contains kanas, all characters in that field is rendered as the Japanese style;
b) otherwise, characters with variants are rendered in the style of the UI language.

(the 置 in the image below are added by me, not in the original metadata)
![Image](https://github.com/user-attachments/assets/80bffdac-1b05-4cd1-98ac-d4e21b223534)
while this is not 100% correct, it does cover the majority of cases (if a song is in Japanese its title is more likely to contain kanas instead of being kanji only), and I doubt you can do significantly better for instrumental songs (unless you know its meaning i.e. niwashi=gardener, I don't think you can tell if 庭師 is Japanese or traditional Chinese because the writing of kanjis are same).

**Comment by soloopooo** [comment]
The font in osu! stable uses the system(Windows) font in default, as:
Microsoft YaHei ( Chinese Simp. )
Microsoft JhengHei (Chinese Trad. )
Meiryo ( Japanese )
Gulim(?) ( Korean )

So, in osu stable, the font glyph is right in regular UI when user set their language correctly. Also in beatmap metadata.

In lazer, the CJK font has 4 variables : JP, KR, ZH-CN, ZH-TW.

Game will use the alphabet order to choose which variable shows in UI, so the default is JP variable.

This is an issue that SOO....many games also have in.

**Comment by Akarinnnnn** [comment]
~~Solution proposal: #36949~~ Not working, thinking a new proposal...

**Comment by Akarinnnnn** [comment]
> I'm 99% sure that isn't the solution we want, as it would display incorrect characters in metadata when the language of the metadata doesn't match the user's language. Maybe broken in a better way than it is now, but still an issue to consider.
> 
> As mentioned by [@bdach](https://github.com/bdach) this should be localised to usages rather than applied on a game-wide level. Maybe something which is added to `LocalisedString`.

I'm thinking about annotate text span with CJK variation info first, like this, if not already:

<img width="968" height="477" alt="Image" src="https://github.com/user-attachments/assets/03e1003b-4136-4ae0-b144-7a4588e56622" />
